File: .local/lib/copilot_workflows/tests/fake_copilot.py
#!/usr/bin/env python3
"""A fake `copilot` for zero-credit testing of the cwf runtime.

It speaks just enough of the real `copilot -p ... --output-format json` contract:
emit some ignorable JSONL events, one `assistant.message` with `content`, and a
terminal `result` carrying `sessionId`, `exitCode`, and a session shutdown AIC record.

Behavior is steered by directives embedded anywhere in the prompt:

    [[FAKE:{"category": "bug"}]]      -> emits that JSON as the message's final line
    [[FAKE:{"_content": "hello"}]]    -> uses "hello" as the whole message content
    [[FAKE:{"_cr": 0.5, ...}]]        -> reports 0.5 fallback cost units
    [[FAKE:{"_exit": 2}]]             -> exits non-zero (simulate failure)
    [[FAKE:{"_sleep": 1.0}]]          -> sleeps 1.0s before output (simulate a hang)
    [[FAKE:{"_session_error": "x"}]]  -> emits a soft session.error event

The LAST directive in the prompt wins. With no directive, it echoes the prompt.
"""
import json
import os
import re
import sys
import time
import uuid


def main() -> int:
    argv = sys.argv[1:]
    prompt = ""
    for i, a in enumerate(argv):
        if a in ("-p", "--prompt") and i + 1 < len(argv):
            prompt = argv[i + 1]
            break

    payload = {}
    for raw in re.findall(r"\[\[FAKE:(.*?)\]\]", prompt, re.S):
        try:
            payload = json.loads(raw)
        except Exception:
            payload = {}

    cr = 0.01
    exit_code = 0
    content = None
    sleep_s = 0.0
    session_error = None
    error_type = "model_call"
    if isinstance(payload, dict):
        cr = float(payload.pop("_cr", 0.01))
        exit_code = int(payload.pop("_exit", 0))
        sleep_s = float(payload.pop("_sleep", 0.0))
        session_error = payload.pop("_session_error", None)
        error_type = str(payload.pop("_error_type", error_type))
        if "_content" in payload:
            content = str(payload.pop("_content"))

    if content is None:
        if payload:
            content = "stub reasoning.\n" + json.dumps(payload)
        else:
            content = "stub reply to: " + prompt[:60].replace("\n", " ")

    sid = str(uuid.uuid4())
    out_tokens = max(1, len(content) // 4)
    copilot_home = os.environ.get("COPILOT_HOME") or os.path.expanduser("~/.copilot")
    state = os.path.join(copilot_home, "session-state", sid)
    os.makedirs(state, exist_ok=True)
    with open(os.path.join(state, "events.jsonl"), "w", encoding="utf-8") as fh:
        fh.write(json.dumps({
            "type": "session.shutdown",
            "data": {"shutdownType": "routine", "totalNanoAiu": int(cr * 1_000_000_000)},
        }) + "\n")

    def emit(obj):
        sys.stdout.write(json.dumps(obj) + "\n")

    if sleep_s > 0:  # simulate a hung agent so the runtime's timeout/kill path can be tested
        time.sleep(sleep_s)

    # ignorable events the parser must skip over
    emit({"type": "session.skills_loaded", "data": {"skills": []}, "ephemeral": True})
    emit({"type": "assistant.turn_start", "data": {"turnId": "0"}})
    emit({"type": "assistant.reasoning_delta", "data": {"deltaContent": "thinking"}, "ephemeral": True})
    if session_error is not None:
        emit({"type": "session.error", "data": {
            "errorType": error_type, "message": str(session_error),
        }})
    emit({"type": "assistant.message", "data": {
        "messageId": sid, "model": "fake", "content": content,
        "toolRequests": [], "outputTokens": out_tokens,
    }})
    emit({"type": "assistant.turn_end", "data": {"turnId": "0"}})
    emit({"type": "result", "sessionId": sid, "exitCode": exit_code, "usage": {
        "premiumRequests": cr, "totalApiDurationMs": 1, "sessionDurationMs": 1,
        "codeChanges": {"linesAdded": 0, "linesRemoved": 0, "filesModified": []},
    }})
    return exit_code


if __name__ == "__main__":
    raise SystemExit(main())

File: .copilot/extensions/workflow/fixtures/fake-copilot.mjs
#!/usr/bin/env node
/**
 * @module fixtures/fake-copilot
 *
 * A zero-cost stand-in for the real `copilot` binary, used by the workflow test suite
 * (`CWF_COPILOT_BIN` points at this file). It mimics `copilot -p <prompt> --output-format json`:
 * it prints a JSONL event stream to stdout (an `assistant.message` then a `result`) and writes a
 * child session log at `$COPILOT_HOME/session-state/<id>/events.jsonl` containing a
 * `session.shutdown` record with `totalNanoAiu` + per-model token usage — the exact shape
 * `agent.mjs` reads back for AIC/token accounting.
 *
 * Behaviour is driven entirely by env vars so tests can exercise every branch without spending
 * real AI credits:
 *
 * - `CWF_FAKE_MODE`   — `ok` (default) | `fail` | `hang` | `treehang` | `malformed` | `nojson` | `sessionerror`
 * - `CWF_FAKE_NANO_AIU` — session-wide nanoAIU to report (default `500000000` = 0.5 AIC)
 * - `CWF_FAKE_OUTPUT_TOKENS` — output tokens to report (default `42`)
 * - `CWF_FAKE_CONTENT` — assistant content (default `ECHO: <prompt>`)
 * - `CWF_FAKE_STDERR` — stderr text for `fail` mode
 * - `CWF_FAKE_SESSION` — force a child session id (default: random)
 * - `CWF_FAKE_DELAY_MS` — delay a successful response (for concurrency tests)
 * - `CWF_FAKE_PID_FILE` — write this fake process's PID to the given path
 * - `COPILOT_HOME`     — where the child session log is written (default `~/.copilot`)
 */
import { spawn } from "node:child_process";
import { appendFileSync, existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { randomBytes } from "node:crypto";

/**
 * Extract the value following `flag` in an argv array, or `undefined`.
 * @param {string[]} argv
 * @param {string} flag
 * @returns {string | undefined}
 */
function argValue(argv, flag) {
	const i = argv.indexOf(flag);
	return i >= 0 && i + 1 < argv.length ? argv[i + 1] : undefined;
}

const argv = process.argv.slice(2);
const prompt = argValue(argv, "-p") ?? "";
const model = argValue(argv, "--model") ?? process.env.CWF_FAKE_MODEL ?? "fake-model";
const mode = process.env.CWF_FAKE_MODE ?? "ok";
const nanoAiu = Number(process.env.CWF_FAKE_NANO_AIU ?? 500_000_000);
const outputTokens = Number(process.env.CWF_FAKE_OUTPUT_TOKENS ?? 42);
const content = process.env.CWF_FAKE_CONTENT ?? `ECHO: ${prompt}`;
const resumedSessionId = argValue(argv, "--resume");
const sessionId = resumedSessionId ?? process.env.CWF_FAKE_SESSION ?? `fake-${randomBytes(6).toString("hex")}`;
const copilotHome = process.env.COPILOT_HOME || join(homedir(), ".copilot");
const delayMs = Math.max(0, Number(process.env.CWF_FAKE_DELAY_MS ?? 0));
if (process.env.CWF_FAKE_PID_FILE) writeFileSync(process.env.CWF_FAKE_PID_FILE, String(process.pid), "utf8");

/**
 * Build one cumulative `session.shutdown` record.
 * @param {number} nano
 * @param {{ inputTokens: number, outputTokens: number, cacheReadTokens: number, cacheWriteTokens: number, reasoningTokens: number }} usage
 */
function mkShutdown(nano, usage) {
	return {
		type: "session.shutdown",
		sessionId,
		data: {
			shutdownType: "normal",
			currentModel: model,
			codeChanges: { filesModified: [], linesAdded: 0, linesRemoved: 0 },
			sessionStartTime: Date.now() - 1000,
			totalApiDurationMs: 123,
			totalNanoAiu: nano,
			modelMetrics: {
				[model]: {
					requests: { count: 1, cost: 1 },
					totalNanoAiu: nano,
					usage,
				},
			},
		},
	};
}

/**
 * Write cumulative child usage. Resumed turns append a new shutdown to the same session log.
 */
function writeShutdownLog() {
	const dir = join(copilotHome, "session-state", sessionId);
	mkdirSync(dir, { recursive: true });
	const path = join(dir, "events.jsonl");
	const previous = readPreviousUsage(path);
	const usage = {
		inputTokens: previous.inputTokens + 100,
		outputTokens: previous.outputTokens + outputTokens,
		cacheReadTokens: previous.cacheReadTokens + 10,
		cacheWriteTokens: previous.cacheWriteTokens + 5,
		reasoningTokens: previous.reasoningTokens + 7,
	};
	const total = previous.nanoAiu + nanoAiu;
	const line = JSON.stringify(mkShutdown(total, usage)) + "\n";
	if (resumedSessionId && existsSync(path)) appendFileSync(path, line, "utf8");
	else writeFileSync(path, line, "utf8");
	if (process.env.CWF_FAKE_NANO_AIU_2) {
		appendFileSync(path, JSON.stringify(mkShutdown(Number(process.env.CWF_FAKE_NANO_AIU_2), usage)) + "\n", "utf8");
	}
}

/** @param {string} path */
function readPreviousUsage(path) {
	const empty = { nanoAiu: 0, inputTokens: 0, outputTokens: 0, cacheReadTokens: 0, cacheWriteTokens: 0, reasoningTokens: 0 };
	if (!existsSync(path)) return empty;
	let latest = null;
	for (const line of readFileSync(path, "utf8").split("\n")) {
		if (!line.trim()) continue;
		try {
			const rec = JSON.parse(line);
			if (rec.type === "session.shutdown" && rec.data?.totalNanoAiu != null) latest = rec;
		} catch {
			// Ignore malformed history in the fake just like the production reader.
		}
	}
	if (!latest) return empty;
	const metric = Object.values(latest.data.modelMetrics || {})[0];
	const usage = metric?.usage || {};
	return {
		nanoAiu: Number(latest.data.totalNanoAiu || 0),
		inputTokens: Number(usage.inputTokens || 0),
		outputTokens: Number(usage.outputTokens || 0),
		cacheReadTokens: Number(usage.cacheReadTokens || 0),
		cacheWriteTokens: Number(usage.cacheWriteTokens || 0),
		reasoningTokens: Number(usage.reasoningTokens || 0),
	};
}

/** @param {object} obj */
const emit = (obj) => process.stdout.write(JSON.stringify(obj) + "\n");

function finishOk() {
	writeShutdownLog();
	emit({ type: "assistant.message", data: { content, outputTokens, model } });
	emit({ type: "result", sessionId });
	process.exit(0);
}

switch (mode) {
	case "fail":
		process.stderr.write(process.env.CWF_FAKE_STDERR ?? "fake-copilot: simulated failure\n");
		process.exit(1);
		break;
	case "hang":
		setInterval(() => {}, 1 << 30); // stay alive until the parent's timeout kills us
		break;
	case "treehang": {
		// The leaf inherits stdout/stderr, reproducing the real failure where killing only `copilot`
		// leaves a tool subprocess alive and keeps the parent's JSONL reader open forever.
		const leaf = spawn(process.execPath, ["-e", "setInterval(() => {}, 1 << 30)"], {
			stdio: ["ignore", "inherit", "inherit"],
		});
		if (process.env.CWF_FAKE_PID_FILE && leaf.pid) writeFileSync(process.env.CWF_FAKE_PID_FILE, String(leaf.pid), "utf8");
		setInterval(() => {}, 1 << 30);
		break;
	}
	case "malformed":
		writeShutdownLog();
		process.stdout.write("not json at all\n");
		emit({ type: "assistant.message", data: { content, outputTokens, model } });
		process.stdout.write("{ truncated json\n");
		emit({ type: "result", sessionId });
		process.exit(0);
		break;
	case "nojson":
		writeShutdownLog();
		process.stdout.write("just some prose, no events\n");
		emit({ type: "result", sessionId });
		process.exit(0);
		break;
	case "sessionerror":
		writeShutdownLog();
		emit({ type: "session.error", data: { errorType: "RateLimit", message: "slow down" } });
		emit({ type: "result", sessionId });
		process.exit(0);
		break;
	case "nousage":
		emit({ type: "assistant.message", data: { content, outputTokens, model } });
		emit({ type: "result", sessionId });
		process.exit(0);
		break;
	case "ok":
	default:
		if (delayMs) setTimeout(finishOk, delayMs);
		else finishOk();
}
